### ISSUE TITLE
sndfile-convert crashes in d2ulaw_array ()

### ISSUE THREAD

**Issue by gsharpsh00ter** [erikd/libsndfile#368]
**0x01 Description**

In libsndfile 1.0.28 and the master branch, a segfault issue was found which allows remote attackers to cause a denial of service or possibly have unspecified other impact via a crafted audio file.

Please confirm.

**0x02 Reproduce**

This issue can be reproduced by the following command:
`sndfile-convert -normalize $POCFILE out.aif`

**0x03 0x03 Cause Analyze**

This issue should have something to do with the "-normalize" option.

sndfile-convert run as expected without "-normalize" option:
`gzq@ubuntu:~/code/tmp/build$ ./sndfile-convert /home/gzq/run/libsndfile/crashes/poc1 out.aif
`
sndfile-convert crashed with "-normalize" option:
`gzq@ubuntu:~/code/tmp/build$ ./sndfile-convert -normalize /home/gzq/run/libsndfile/crashes/poc1 out.aif
`Segmentation fault

Debug information is as follow:

```
gzq@ubuntu:~/code/tmp/build$ gdb -q ./sndfile-convert
Reading symbols from ./sndfile-convert...done.
(gdb) r -normalize /home/gzq/run/libsndfile/crashes/poc1 out.aif
Starting program: /home/gzq/code/tmp/build/sndfile-convert -normalize /home/gzq/run/libsndfile/crashes/poc1 out.aif

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7b5123f in d2ulaw_array (ptr=0x555555759280 <data>, count=0, 
    buffer=0x7fffffffc2e0 "\200\200\200\200\227tt\214\214\214\227tt\200\200\200\227\214\214\227\200t\200\200\214\214\200\200\200i\200\227\214ti\200\200\200tt\200t\200\200iti\200\200ttitt\214i]it]ii]\214t\200\200titt\200\214\200t\227\214\200t\214\214\214\200\200\200\243\214\214i]ii]\227\200tiiti\200t\200t\214\200\200t\214\243\200t\200\214t\200\200t\214\227iii\214\200]]ti\214t\214\227tt\200\227iit\214t]i\200i\200\214Qt\214]itt\214]itt\200\214it\214\227\200i\227\243\200\214t\200\227Q\214t\214i\200\200tt\227\200t\227\200]i\214F]\214]]i\243i"..., normfact=8191.75) at /home/gzq/code/tmp/src/ulaw.c:853
853				buffer [count] = 0x7F & ulaw_encode [- lrint (normfact * ptr [count])] ;
(gdb) bt
#0  0x00007ffff7b5123f in d2ulaw_array (ptr=0x555555759280 <data>, count=0, 
    buffer=0x7fffffffc2e0 "\200\200\200\200\227tt\214\214\214\227tt\200\200\200\227\214\214\227\200t\200\200\214\214\200\200\200i\200\227\214ti\200\200\200tt\200t\200\200iti\200\200ttitt\214i]it]ii]\214t\200\200titt\200\214\200t\227\214\200t\214\214\214\200\200\200\243\214\214i]ii]\227\200tiiti\200t\200t\214\200\200t\214\243\200t\200\214t\200\200t\214\227iii\214\200]]ti\214t\214\227tt\200\227iit\214t]i\200i\200\214Qt\214]itt\214]itt\200\214it\214\227\200i\227\243\200\214t\200\227Q\214t\214i\200\200tt\227\200t\227\200]i\214F]\214]]i\243i"..., normfact=8191.75) at /home/gzq/code/tmp/src/ulaw.c:853
#1  0x00007ffff7b51bb2 in ulaw_write_d2ulaw (psf=0x555555769ca0, ptr=0x555555759280 <data>, len=1) at /home/gzq/code/tmp/src/ulaw.c:1041
#2  0x00007ffff7b6e2f0 in sf_writef_double (sndfile=0x555555769ca0, ptr=0x555555759280 <data>, frames=1) at /home/gzq/code/tmp/src/sndfile.c:2553
#3  0x0000555555556454 in sfe_copy_data_fp (outfile=0x555555769ca0, infile=0x555555766010, channels=1, normalize=1) at /home/gzq/code/tmp/programs/common.c:71
#4  0x0000555555556089 in main (argc=4, argv=0x7fffffffe508) at /home/gzq/code/tmp/programs/sndfile-convert.c:350
```

Enviroment:

```
gzq@ubuntu:~/fuzz/run/libsndfile/crashes$ uname -a
Linux ubuntu 4.8.0-59-generic #64-Ubuntu SMP Thu Jun 29 19:38:34 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
gzq@ubuntu:~/fuzz/run/libsndfile/crashes$ cat /etc/*lease
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.10
DISTRIB_CODENAME=yakkety
DISTRIB_DESCRIPTION="Ubuntu 16.10"
NAME="Ubuntu Kylin"
VERSION="16.10 (Yakkety Yak)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu Kylin 16.10"
VERSION_ID="16.10"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="http://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=yakkety
UBUNTU_CODENAME=yakkety

```

**0x04 Author**

This issue is reported by Ziqiang Gu from WeiRan Labs

**0x05 POCFILE**

[poc-libsndfile-normalize-segfault-1.zip](https://github.com/erikd/libsndfile/files/1800110/poc-libsndfile-normalize-segfault-1.zip)

**Comment by evpobr** [comment]
```
========================================
File : c:\Users\evpob\Source\Repos\libsndfile\CMakeBuild\Debug\segfault.aiff
Length : 1837
RIFF : 1829
WAVE
fmt  : 16
  Format        : 0x7 => WAVE_FORMAT_MULAW
  Channels      : 1
  Sample Rate   : 11025
  Block Align   : 1
  Bit Width     : 8
  Bytes/sec     : 11025
*** 'data' chunk should be an even number of bytes in length.
data : 1
  Unknown chunk marker at position 46. Resynching.
*** Chunk size 2153611369 > file length 1837. Exiting parser.
**** All non-PCM format files should have a 'fact' chunk.

----------------------------------------
Sample Rate : 11025
Frames      : 1
Channels    : 1
Format      : 0x00010010
Sections    : 1
Seekable    : TRUE
Duration    : 00:00:00.000
Signal Max  : 0 (-inf dB)
```

**Comment by evpobr** [comment]
@erikd, it seems like data chunk is incorrect, but libsndfile ignores fact that critical data content is corrupted. I think this file must be rejected at `wav_read_header()` stage.

And i cannot reproduce this on Win32 build, file is opened as WAV file with 1 frame and converted to AIFF successfully.

**Comment by hlef** [comment]
Might be another duplicate of #344, fixed in 8ddc442d539ca775d80cdbc7af17a718634a743f

**Comment by erikd** [comment]
Cannot be reproduced with git HEAD (bef2abc9e888142203953addc31c50a192e496e5).
```
> programs/sndfile-convert -normalize poc-libsndfile-normalize-segfault-1 out.aif
> programs/sndfile-info out.aif 
========================================
File : out.aif
Length : 74
FORM : 66
 AIFC
 FVER : 4
 COMM : 24
  Sample Rate : 11025
  Frames      : 2
  Channels    : 1
  Sample Size : 8
  Encoding    : ulaw
 SSND : 10
  Offset     : 0
  Block Size : 0

----------------------------------------
Sample Rate : 11025
Frames      : 2
Channels    : 1
Format      : 0x00020010
Sections    : 1
Seekable    : TRUE
Duration    : 00:00:00.000
Signal Max  : 32124 (-0.17 dB)
```
